pkg/proxy/handler_test.go

```diff
@@ -1,69 +1,157 @@
 	"github.com/prometheus/client_golang/prometheus/testutil"
 	io_prometheus_client "github.com/prometheus/client_model/go"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"github.com/substratusai/lingo/pkg/deployments"
-	"k8s.io/apimachinery/pkg/runtime"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
-	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
-
-	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/cache"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/config"
-	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
-func TestMetrics(t *testing.T) {
+func TestHandler(t *testing.T) {
+	const (
+		model1 = "model1"
+		model2 = "model2"
+
+		maxRetries = 3
+	)
+	models := map[string]string{
+		model1: "deploy1",
+		model2: "deploy2",
+	}
+
 	specs := map[string]struct {
-		request   *http.Request
-		expCode   int
-		expLabels map[string]string
+		reqMethod string
+		reqPath   string
+		reqBody   string
+
+		backendCode int
+		backendBody string
+
+		expCode                int
+		expBody                string
+		expLabels              map[string]string
+		expBackendRequestCount int
 	}{
-		"with model name": {
-			request: httptest.NewRequest(http.MethodGet, "/", strings.NewReader(`{"model":"my_model"}`)),
-			expCode: http.StatusNotFound,
+		"no model": {
+			reqMethod: http.MethodPost,
+			reqPath:   "/",
+			reqBody:   "{}",
+			expCode:   http.StatusBadRequest,
+			expBody:   `{"error":"unable to parse model: no model specified"}` + "\n",
 			expLabels: map[string]string{
-				"model":       "my_model",
+				"model":       "",
+				"status_code": "400",
+			},
+			expBackendRequestCount: 0,
+		},
+		"model not found": {
+			reqMethod: http.MethodPost,
+			reqPath:   "/",
+			reqBody:   `{"model":"does-not-exist"}`,
+			expCode:   http.StatusNotFound,
+			expBody:   `{"error":"model not found: does-not-exist"}` + "\n",
+			expLabels: map[string]string{
+				"model":       "does-not-exist",
 				"status_code": "404",
 			},
+			expBackendRequestCount: 0,
 		},
-		"empty model name": {
-			request: httptest.NewRequest(http.MethodGet, "/", strings.NewReader("{}")),
-			expCode: http.StatusBadRequest,
+		"happy 200": {
+			reqMethod:   http.MethodPost,
+			reqPath:     "/",
+			reqBody:     fmt.Sprintf(`{"model":%q}`, model1),
+			backendCode: http.StatusOK,
+			backendBody: `{"result":"ok"}`,
+			expCode:     http.StatusOK,
+			expBody:     `{"result":"ok"}`,
 			expLabels: map[string]string{
-				"model":       "",
+				"model":       model1,
+				"status_code": "200",
+			},
+			expBackendRequestCount: 1,
+		},
+		"retryable 500": {
+			reqMethod:   http.MethodPost,
+			reqPath:     "/",
+			reqBody:     fmt.Sprintf(`{"model":%q}`, model1),
+			backendCode: http.StatusInternalServerError,
+			backendBody: `{"err":"oh no!"}`,
+			expCode:     http.StatusBadGateway,
+			expBody:     `{"error":"Bad Gateway"}` + "\n",
```

Comment:
When retry is enabled, the original backend error msg is never passed to the client. This can be a valid design decision but we should be sensible what error codes are "retryable".  Do you have some scenarios for `500` in mind that can succeed on a retry?

Reply:
Good catch, we should probably only filter 500s

Reply:
Scratch that, I am going to just pipe through the response in the case where it got an http response by filtering out retry errors.

Reply:
Updated